tests/phpunit/integration/Core/Authentication/Clients/OAuth_ClientTest.php

```diff
@@ -644,20 +645,54 @@ function ( Request $request ) {
 
 		// Verify the redirect URL was stored in transients.
 		$stored_redirect = $transients->get( $code_hash );
-		$this->assertNotEmpty( $stored_redirect, 'Redirect URL was not stored in transients' );
-		$this->assertEquals( $redirect_url, $stored_redirect );
+		$this->assertEquals( $redirect_url, $stored_redirect, 'The redirect URL should be stored in transients with the correct hash' );
 
-		// Verify the key behavior: that the transient is stored with the correct hash and has the right value.
-		$this->assertEquals(
-			$redirect_url,
-			$transients->get( $code_hash ),
-			'The redirect URL should be stored in transients with the correct hash'
+		// SECOND AUTHORIZATION ATTEMPT.
+
+		// Create a new client instance to test the second attempt.
+		$client2 = new OAuth_Client( $context, null, $user_options, null, null, null, null, $transients );
+		$client2->get_authentication_url( $success_redirect );
+
+		// Intercept the redirect to verify the URL is the same as the first attempt.
+		add_filter(
+			'wp_redirect',
+			function ( $location ) use ( $stored_redirect ) {
+				// When `wp_safe_redirect` is called, check the location matches our transient.
+				$this->assertEquals( $stored_redirect, $location, 'Second call should redirect to stored URL' );
+
+				// Throw a RedirectException to simulate the behavior of `wp_redirect`.
+				throw new RedirectException( $location );
+			},
+			10,
+			1
 		);
 
-		// For the early exit path, verify the key pieces are in place:
-		$this->assertEquals( md5( 'test-code' ), $code_hash, 'Code is properly hashed' );
-		$this->assertNotEmpty( $stored_redirect, 'Redirect URL is stored for later use' );
-		$this->assertEquals( $redirect_url, $stored_redirect, 'Stored redirect matches expected URL' );
+		// For the second client, we can verify it never calls `fetchAccessTokenWithAuthCode`
+		// by setting up a mock that will fail the test if called.
+		$google_client_mock2 = $this->getMockBuilder( 'Google\Site_Kit\Core\Authentication\Clients\Google_Site_Kit_Client' )
+			->setMethods( array( 'fetchAccessTokenWithAuthCode' ) )->getMock();
+
+		$google_client_mock2->expects( $this->never() )
+			->method( 'fetchAccessTokenWithAuthCode' );
+
+		$this->force_set_property( $client2, 'google_client', $google_client_mock2 );
+
+		// Now call `authorize_user` a second time.
+		try {
+			$client2->authorize_user();
+			$this->fail( 'Expected to throw a RedirectException on second attempt!' );
+		} catch ( RedirectException $redirect2 ) {
+			// Our filter above already verified the URL matches, but we should
+			// explicitly assert something here to avoid empty catch block errors.
+			$this->assertInstanceOf(
+				RedirectException::class,
+				$redirect2,
+				'Second call should throw a RedirectException'
+			);
+		} finally {
+			// Clean up our filter.
+			remove_all_filters( 'wp_redirect' );
+		}
```

Comment:
As per my comment above, we can simplify the test and make the assertion in the catch block more useful and consistent:

```suggestion
			// Verify the stored URL is used for the redirect.
			$this->assertEquals( $stored_redirect, $redirect->get_location() );
```

Reply:
Updated.